Rocky.class.nut

```diff
@@ -199,50 +245,84 @@ class Rocky {
         if (contentType.find("multipart/form-data;") == 0) {
             local parts = [];
 
-            // _parse_body is wrapped in a try/catch.. so we just let this fail
-            // when the content-type isn't long enough (and on other issues).
-            local boundary = contentType.slice(30);
+            // Find the boundary in the contentType
+            local boundary;
+            local findString = regexp(@"boundary([ ]*)=");
+            local match = findString.search(contentType);
+
+            if (match) {
+                boundary = contentType.slice(match.end);
+                boundary = strip(boundary);
+            } else {
+                throw "No boundary found in content-type";
+            }
+
+            // Remove all carriage returns from string (to support either \r\n or \n for linebreaks)
+            local body = "";
+            local bodyLines = split(req.body, "\r");
+            foreach (i, line in bodyLines) {
+                body += line;
+            }
 
             local bindex = -1;
             do {
-                bindex = req.body.find("--" + boundary + "\r\n", bindex+1);
+                bindex = body.find("--" + boundary + "\n", bindex+1);
 
                 if (bindex != null) {
                     // Locate all the parts
-                    local hstart = bindex + boundary.len() + 4;
-                    local nstart = req.body.find("name=\"", hstart) + 6;
-                    local nfinish = req.body.find("\"", nstart);
-                    local fnstart = req.body.find("filename=\"", hstart) + 10;
-                    local fnfinish = req.body.find("\"", fnstart);
-                    local bstart = req.body.find("\r\n\r\n", hstart) + 4;
-                    local fstart = req.body.find("\r\n--" + boundary, bstart);
-
-                    // Pull out the parts as strings
-                    local headers = req.body.slice(hstart, bstart);
+                    local hstart = bindex + boundary.len() + 3;
+                    local hfinish = body.find("\n\n", hstart);
+                    local header = body.slice(hstart, hfinish);
+
+                    // Get the name
                     local name = null;
+                    local nstart = header.find(" name=\"");
+                    if (nstart != null) {
+                        nstart += 7;
+                        local nfinish = header.find("\"", nstart);
+                        name = header.slice(nstart, nfinish);
+                    } else {
+                        throw ERROR_MISSING_NAME;
+                    }
+
+                    // Get the filename
                     local filename = null;
+                    local fnstart = header.find(" filename=\"");
```

Comment:
If I have to leave this code (I won't probably do so), I would make it:
```
local filename = null;
local searchString = " filename=\"";
if (fnstart != null) {
    local fnfinish = header.find("\"", fnstart + searchString.len());
    filename = header.slice(fnstart, fnfinish);
}
```

Reply:
I mean I would just avoid hardcoding the length of the strings.

Reply:
Yeah, the hardcoding the string lengths was already there in the code, I probably should have changed that. I believe I should still search for "filename=" like this:
`local fnstart = header.find(" filename=\"");`
rather than this:
`local fnfinish = header.find("\"", fnstart + searchString.len());`
because the order of the parameters in the body isn't fixed.

Reply:
I guess this should be taken out if we substitute internal loop parser with a reg-exp